File: challenges/2026-09-20-high-score-reel-combinations/solution.java
import java.util.*;

public class solution {

    public static List<String> highScoreCombinations(List<List<Character>> reels, List<String> bannedPrefixes) {
        // TODO: implement
        return new ArrayList<>();
    }

    public static void main(String[] args) {
        List<Object[]> tests = new ArrayList<>();
        tests.add(new Object[]{
            "example 1",
            List.of(List.of('A', 'B'), List.of('C', 'D'), List.of('E', 'F')),
            List.of("BD"),
            List.of("ACE", "ACF", "ADE", "ADF", "BCE", "BCF")
        });
        tests.add(new Object[]{
            "example 2",
            List.of(List.of('X'), List.of('Y', 'Z'), List.of('A', 'B', 'C')),
            List.of("XZ", "XYB"),
            List.of("XYA", "XYC")
        });
        tests.add(new Object[]{
            "example 3",
            List.of(List.of('A'), List.of('B'), List.of('C')),
            List.<String>of(),
            List.of("ABC")
        });
        tests.add(new Object[]{
            "edge: first letter banned prunes everything",
            List.of(List.of('A'), List.of('B'), List.of('C')),
            List.of("A"),
            List.<String>of()
        });
        tests.add(new Object[]{
            "edge: mid-level banned prefix prunes a branch",
            List.of(List.of('P', 'Q'), List.of('R'), List.of('S', 'T')),
            List.of("QR"),
            List.of("PRS", "PRT")
        });

        for (Object[] test : tests) {
            String name = (String) test[0];
            @SuppressWarnings("unchecked")
            List<List<Character>> reels = (List<List<Character>>) test[1];
            @SuppressWarnings("unchecked")
            List<String> banned = (List<String>) test[2];
            @SuppressWarnings("unchecked")
            List<String> expected = (List<String>) test[3];
            List<String> actual = highScoreCombinations(reels, banned);
            String status = actual.equals(expected) ? "PASS" : "FAIL";
            System.out.println("[" + status + "] " + name + ": expected=" + expected + " actual=" + actual);
        }
    }
}

File: challenges/2026-09-20-high-score-reel-combinations/solution.ts
function highScoreCombinations(reels: string[][], bannedPrefixes: string[]): string[] {
  // TODO: implement
  return [];
}

function runSelfChecks(): void {
  const cases: [string, string[][], string[], string[]][] = [
    [
      "example 1",
      [["A", "B"], ["C", "D"], ["E", "F"]],
      ["BD"],
      ["ACE", "ACF", "ADE", "ADF", "BCE", "BCF"],
    ],
    [
      "example 2",
      [["X"], ["Y", "Z"], ["A", "B", "C"]],
      ["XZ", "XYB"],
      ["XYA", "XYC"],
    ],
    ["example 3", [["A"], ["B"], ["C"]], [], ["ABC"]],
    [
      "edge: first letter banned prunes everything",
      [["A"], ["B"], ["C"]],
      ["A"],
      [],
    ],
    [
      "edge: mid-level banned prefix prunes a branch",
      [["P", "Q"], ["R"], ["S", "T"]],
      ["QR"],
      ["PRS", "PRT"],
    ],
  ];

  for (const [name, reels, bannedPrefixes, expected] of cases) {
    const actual = highScoreCombinations(reels, bannedPrefixes);
    const status =
      actual.length === expected.length && actual.every((v, i) => v === expected[i])
        ? "PASS"
        : "FAIL";
    console.log(`[${status}] ${name}: expected=${JSON.stringify(expected)} actual=${JSON.stringify(actual)}`);
  }
}

runSelfChecks();
